Basic SAVE routine

dikdom opened the pull request:
Sorry for the lot's of 'noise' in the commit messages. I was implementing on my 'save' branch, and merging from/to master caused a lot of unwanted commit messages.
- Basic SAVE implemented
- Overwrite implemented (e.g. SAVE"@:existing file",8,1)
- No error message on SAVE failure (file exists, not enough space, IO error, etc...)
- No error message on ERROR CHANNEL
- I didn't find any difference between ,8,1 and ,8, so both of them yields the same. (though there may be some, please let me know!)
- Only .d64 file support is implemented (.d74 and .d81 support is missing)
- File as REL cannot be saved, only prg and seq types can be saved (SEQ is saved just as if it was a PRG - according to emulators and KERNAL)

KimJorgensen commented:
First of all, thank you for this pull-request. This feature will make a nice addition to Kung Fu Flash.
As you might expect, a lot of decisions has been made in the existing code base, some of which may not be obvious so I have some comments about how to better align with the existing code.

I’ll add these comments to the source code but there is one big thing that I would like to address generally and that is the addition of D64 write support:

What I would prefer is an extension of d64_reader (renamed) with a similar set of function for writing to disk. No offense, but the d64_writer feels like it is bolted on the existing design, with a different “API”, state handling and partially duplicated code (sector handling).
If I understood you correctly, part of the reason for this different design was to postpone writing the updated BAM and directory to disk until a file has been successfully saved. However, I’m not sure why you think that this was important as the original 1541 drive does to my knowledge not behave this way?

dikdom commented:
With the d64_writer I was indeed in a problem: i couldn't separate the 'write' functions as I wanted. And as time passed I realized that the writing itself is nothing more than write_sector(). 
The development evolved from direct sector writing (even directory and BAM) to memory buffer writes and sync at the end.
The reason I choosed to buffer
- BAM: there are several changes on it during write: possible dir entry allocation, sector allocation one-by-one and I wanted to spare the SD from the multiple writes
- dir sectors: I wanted to finalize the save when the data is already on the disk so no ruined directory can be resulted on a failed disk write
I know these are deviate from the original 1541 routines, but here we have a different data storage media also.

KimJorgensen commented:
* BAM: It makes sense to spare the SD card from writes but due to the limited cache in FatFs you will still get a write to the SD card for each D64 sector written so I'm not sure how much this BAM buffer really matters in practice. If a buffer is used it should be part of the existing state struct (D64) and not something the caller should be concerned with. This will allow easier code reuse e.g. from menu_d64.
* dir sectors: I'm not sure why you want to make a less accurate emulation here. Isn't it trivial to "recover" from a partial written file, something you also would be facing when implementing support for ICKOUT? Or maybe I don't fully understand the 'ruined directory' scenario you are talking about.

dikdom commented:
Hello Kim,
actually I'm not that familiar with the 1541 ROM routines. I don't know if the BAM is being continuously written during a SAVE or just after the save is successful. I could image that in the 1541 the BAM is written only at the end. Nevertheless, what I do is the following (A):
- 0th phase: check if file size fits. If not, exit.
- 1st phase: create a new directory entry in its final form (set file size, type is 0x82, first track-sector is also set)
  - if it is an overwrite and file exists indeed then the BAM needs to be freed for the previous file (following the t-s chain except the first t-s pair), then set the file-size. The first T-S is already given and remains allocated - that will be used for sure.
  - look for the first DEL entry type (type && 0x07 == 00). 
    - If the last sector containing the directory entries is full, then a new sector has to be allocated in BAM, this sector must be linked from the last one. This is the only case when there is two directory sector is cached. I allocate and create the new dir entry.
    - If not, then I simply create the dir entry.
- 2nd phase: the firmware is receiving the 0x90 bytes long packages (in the first round there are two extra bytes for the start address). Each time a sector is full, a new one has to be allocated and the new T-S pair is written to the beginning of the full one, then this sector is sent to FatFs to write. Last sector has no next T-S, instead it has a 0-(num of bytes + 1) pair written to the first 2 bytes of the sector
- 3rd phase: write the BAM and the directory sectors. (one is optional and is written only if a new directory sector is allocated as the old sector is changed, because of the linking bytes).
The FatFs has some cache, but as far as I know it is 512 bytes. I think it is continuous write only, but that is just an educated guess.

Instead of this I think the 1541 ROM would do this (B):
- creates the file entry with intermediate file type (0x02 - "PRG<" ?) and set (leave) file size to 0. Allocate a new sector for directory - if necessary and that is also written to BAM.
- loads the BAM and uses this as cache (I'm not sure, but I think there is no continuous seeking to track 18 and back after a sector is written)
- writes the data to the proper sectors and updates the cached BAM - if there is no error/there is space left
- writes the BAM to the disk and updates the file type to 0x82 and set file size to the actual.

The dir sector (because of the dir entry's type field) is written twice, the BAM is written twice ( I don't know if the whole sector is rewritten in case of the 1541)

What I've changed:
- no disk full during save, the write will not start if the program doesn't fit.
- the BAM is written only once (BAM write after new directory sector allocation is spared)
- the dir entry is written only once causing only a properly written file will have a directory entry (there won't be floating directory entry..)
- the SAVE is not stoppable by R/S key

The FatFS cache is 512 bytes (probably continuous) long if I'm not mistaken. One sector fits, but because the sectors are not written continuously then only one sector can fit. I think. 

Long story short: I wanted to buffer as much as possible, to avoid double writes. 
I agree that BAM should be buffered in D64 structure - I just didn't want to touch existing structures. Then during the overwrite implementation I needed the last visited sector - where did the find_file find the directory entry. So I changed D64 as well.. 
Dir sectors: still, I just wanted to avoid double writes. As creating a new dir sector would require to write to BAM, I buffered the dir sectors too.

So I accept: let's change the save process as (B) but I would keep the disk free check at the beginning. The BAM will be written twice. Each SAVE would update the BAM twice. On the SD card, on the same place. This sounds a bit scary to me. What do you think about this?
On the other hand not only BAM needs to cached, but the directory entry also, that needs to be updated. Or shall the routine find at the end of the save routine again?

dikdom commented on firmware/d64_reader.h:
```diff
@@ -144,6 +147,8 @@ typedef struct
     char *diskname;         // valid after d64_read_disk_header
     D64_DIR_ENTRY *entry;   // valid after d64_read_dir
+    uint8_t last_read_track;
+    uint8_t last_read_sector;
```
These were introduced because of the find_file should have returned the directory sector also (track is not that important at D64, but at D81...)

dikdom commented on firmware/d64_reader.h:
```diff
@@ -23,6 +23,9 @@
  */
 
+#ifndef D64_READER_H
+#define D64_READER_H
```
I didn't understand why you made the whole source within one file (include "xxx.c" - I have never seen this kind of include before... )

dikdom commented on firmware/commands.h:
```diff
@@ -84,6 +85,9 @@ typedef enum {
     REPLY_END_OF_FILE,
+    REPLY_NOT_SUPPORTED,
+    REPLY_SAVE_OK,
+    REPLY_SAVE_NOK,
```
Sure. Just a label.. :)
Have you seen how I left the result codes on the varous error scenarios? I wanted them all SAVE_NOK, but during the debug phase they became 1, 2, 3, 4, ... etc...

dikdom commented on firmware/d64_writer.c:
```diff
@@ -0,0 +1,307 @@
+#include "d64_reader.h"
+#include "d64_writer.h"
+
+bool allocate_if_free(D64_SAVE_BUFFER *save_buffer, uint8_t track, uint8_t sector) {
```
I thought static keyword for a function is simply modifies the visibility of that particular function: it is visible only within that compilation unit. But as everything is included at the end, everything is visible so I thought I could leave the static.
I can add it, sure.

KimJorgensen replied:
Yes, the static keyword does modify the visibility and the code will still compile without it. However, telling the compiler that a function is only used within the compilation unit will allow it to perform optimizations such as inlining or register allocation which might not be possible for a function with external linkage.

dikdom commented on firmware/disk_drive.c:
```diff
@@ -200,6 +201,12 @@ static void disk_parse_filename(char *filename, PARSED_FILENAME *parsed)
     char *f_ptr = filename;
     char c, last_c = 0;
     parsed->drive = 0;
+    if(*f_ptr == '@') {
```
true. I tried to be fix all of the (I came from java world...), but surely I missed a few of them.

dikdom commented on launcher/disk.s:
```diff
@@ -825,17 +1038,176 @@ load_next_bank:
+        jsr send_filename
+
+        lda #$00
+        sta $02
```
Correct, debug code..

dikdom commented on firmware/disk_drive.c:
```diff
@@ -466,6 +473,196 @@ static void disk_handle_load(D64 *d64, char *filename, uint8_t *ptr)
+    
+    if(!d64_read_disk_header(d64)) 
+    {
+        c64_send_reply(1);
```
Actually these shall be simply SAVE_NOK (SAVE_ERROR). The C64 cannot do anything with the result sub-value. Just not fine is enough. The error channel should somehow contain the relevant info..

dikdom commented on launcher/disk.s:
```diff
@@ -28,9 +28,11 @@
 CINT	        = $ff81                 ; Initialize screen editor
 RESTOR          = $ff8a                 ; Restore I/O vectors
 BSOUT           = $ffd2                 ; Write byte to default output
+RESTORE_ST_ADDR = $fb8e                 ; STAH->SAH ; STAL-> SAL
```
true. Was used, not anymore.

dikdom commented on launcher/disk.s:
```diff
@@ -92,12 +100,20 @@ CMD_CLOSE               = $83
 CMD_UNTALK              = $85
 CMD_GET_CHAR            = $86
+CMD_SAVE                = $87
+CMD_PUT_BYTE            = $88
```
true.
(PUT_BYTE?? I thought I've deleted.)

KimJorgensen commented:
Hello,
Looking at specs, it seems that most SD cards can withstand around 10,000 full rewrites and thanks to wear leveling this number is much higher for rewriting a few individual files. So I think you are putting too much emphasis on trying to avoid extra writes for no practical benefit. Sure, I would like the design to be efficient and not perform unnecessary writes, but I do also care about reducing code complexity and keeping things consistent.

Regarding the original 1541 behavior, I know when overwriting a file that the new one is saved first and then the space is freed for the old one. However, I haven't studied the details of the 1541 ROM to see how the sector handling is done or what data is cached. That could be an interesting exercise to ensure emulation accuracy and could maybe give some ideas for improving the implementation.

Then there is the disk free check which I’m really not sure why you want. From my perspective it is strictly worse to add this, as this increases the complexity (having to calculate, send, and check the file size upfront) and decreases the emulation accuracy.

dikdom commented on launcher/disk.s:
```diff
@@ -484,7 +674,10 @@ kff_open:
         clc
 @open_done:
-        jmp all_done
+        jsr all_done
```
True, remained there for historical reasons...

dikdom commented on launcher/disk.s:
```diff
@@ -571,9 +765,12 @@ kff_chkin:
         clc                             ; OK
-        jmp all_done
+        jsr all_done
+        jsr copy_disable_ef_rom
```
As a matter of fact this copy is not needed here at all. The 'disable_ef_rom' routine is must be in place when a routine exists. So when a routine enters it is still there and remains there unless that particular routine doesn't modify the RAM (at that place). 
I will fix this and check the other occurrences too.

dikdom commented on firmware/disk_drive.c:
```diff
@@ -466,6 +473,196 @@ static void disk_handle_load(D64 *d64, char *filename, uint8_t *ptr)
+    }
+   
+    disk_parse_filename(filename, &parsed_filename);
+    if (parsed_filename.drive!=0)               // this is a 'single drive system'
```
I'm a bit confused here. I thought in the 'd:filename' notation the 'd' meant the drive number, in case of a dual drive system. Hence 'd' can be either 0 or 1 - KFF and 1541 defaults to '0'.
In case 1 is provided KFF drops the request and forwards it to the IEC bus. I think either SD2IEC nor a real 1541 could handle that request.

KimJorgensen replied:
SD2IEC, Pi1541 and apparently CMD drives uses drive numbers for accessing multiple partitions. The default partition is 1, so even if you only have one partition you can still access the content of the SD card using drive number 0 or 1. The SD2IEC does also support a special partition for hardware configurations.
So a SD2IEC can handle a request to drive number 1, or higher depending on the number of partitions. A real 1541 would only respond to drive number 0.

dikdom commented on launcher/disk.s:
```diff
@@ -234,8 +262,32 @@ basic_mount_starter:
+        lda USB_DATA
+        rts        
+.reloc
+ef3usb_send_receive_rom_addr_end:
```
I think the wait_usb_tx shall not be separated from its LDA USB_DATA pair, that is not nice.. Anyway, that would be 5 bytes.
The last usb rx could also be moved to the caller (3 more bytes in the modified code). Altogether 8 bytes save in RAM. Would it worth?

KimJorgensen replied:
No it is not nice but it really depends on how much RAM you are willing to use here. If it should be super compatible, you could also make a back-up of the RAM first and then restored it after use. Then size matters less (from a compatibility perspective) unless it collides with the stack pointer.
Regarding the last usb rx; you can just make the code "fall through" so no need to add extra instructions to call the subroutine.

dikdom replied:
I replaced the inlined wait and LDA USB_DATA with a subroutine call (JSR) and the size became around $10 less. 
"fall through" just realized what it means. Nice, I will do that.

dikdom commented:
First batch of changes. 
The big part is still ahead of me: 
- the major refactor of the save procedure (including d64_writer.c)
- save protocol change between c64 and MCU
- renaming blocks in EF_RAM
Reading a file byte-by-byte already takes the same amount of time as v1.13.

dikdom commented on launcher/disk.s:
```diff
@@ -340,15 +415,12 @@ enable_ef_rom:
-        rts
+kernal_call_end = *
+.reloc
+disk_api_resident_end:
```
The block is resident in the sense it must be in the RAM when the BASIC interpreter is running. During the KFF execution its content is repurposed temporarily, but when the KFF IO routine exists the resident code is again in its place.

dikdom commented on firmware/disk_drive.c:
```diff
@@ -602,6 +799,10 @@ static void disk_loop(void)
     D64 *d64 = &d64_state.d64;  // Reuse memory from menu
     DISK_CHANNEL *channel = 0, *channels = (DISK_CHANNEL *)crt_ram_banks[1];
+    D64_SAVE_BUFFER *save_buffer = 
+        (D64_SAVE_BUFFER *) (((uint8_t *)crt_ram_banks[1]) + sizeof(DISK_CHANNEL)*16);
```
D64_SAVE_BUFFER will be refactored and will accommodate the following information: 
- buffered BAM (the real BAM is updated at the end of save)
- number of allocated blocks (this will be the file size)
- buffered sector data (to be written to disk)
- valid bytes in this sector data
- T-S position of the buffered sector (where to write, when full/save done)

New structure... It will be needed anyways if you'd like to emulate the 1541 buffers (#) also, later.. But I would like to leave that refactor to you when the time comes.

dikdom commented on firmware/disk_drive.c:
```diff
@@ -466,6 +473,196 @@ static void disk_handle_load(D64 *d64, char *filename, uint8_t *ptr)
+    save_buffer->data_sector.d64sector.data[1] = start_address / 256;
+    save_buffer->data_sector.pos = 2;
+
+    c64_send_reply(REPLY_SAVE_OK);
```
I am still reluctant to put the start address in the first data buffer as it makes pretty complicated in the C64 side to build the buffer properly cycle-by-cycle (the first round is very different), and a the routine that should be executed in EF RAM shall be simple and compact.

dikdom commented:
I have a few issues to discuss, I would like to hear your opinion each of the items below:
- I plan to have a new CMD_SAVE_BUFFER command and the first CMD_SAVE_PRG will have the start address also. This one will create the directory entry, fills the buffers, etc.. CMD_SAVE_BUFFER is called when the EF_RAM buffer is filled by the C64 and can be written to disk (second, cyclic phase of the SAVE prg procedure). What do you think?
- d64_writer.c will contain only those methods that are directly linked to the SD card writing: write_sector(d64, sector_data, t,s), write_buffer(d64, buffer, t, s, offset), flush(d64). Is it what you thought?
- holding the data for a sector to be written (during SAVE) can be the D64_SECTOR in the D64 structure. Holding the BAM during the save needs a separated, dedicated memory area. In the 1541, channel #1 is used for SAVE somehow, so we could use the D64_SECTOR of channel #1 (close it at the beginning of save). Do you want that level of compatibility?
- Overwrite: I think I have a reason why C= engineers didn't implement it in my way. So, during overwrite the user could press R/S key, and the save stopped. I think (didn't try it yet) the 'verify disk' command could revert the pointer in the directory entry to the head of the original T-S series, reverting the original file. In our case it doesn't make much sense as it takes a lot less time to overwrite a file... Are you sure you would like to have it that way? Do you think that I should implement the handling of the R/S key also? (I'm afraid it is handled in the IRQ routine of C64, so direct keyboard scan should be involved)
- I plan to set the ST ($90) variable according to the result of the SAVE. Something to check until there is no real information on the command channel.

dikdom commented on firmware/disk_drive.c:
```diff
@@ -466,6 +473,196 @@ static void disk_handle_load(D64 *d64, char *filename, uint8_t *ptr)
+
+    c64_send_reply(REPLY_SAVE_OK);
+    uint8_t final_reply = 0;
+    if(d64_receive_save_data_cycle(d64, save_buffer))
```
Plan to refactor SAVE, see below.

dikdom commented:
I think yesterday I found a bug in the 1541 ROM. So, when saving a file that doesn't fit, of course the ST will be set (to -125 or 131) and the disk-write led blinks. But loading the directory it says there is no free space. After initializing the disk (or removing and putting it back) and loading the directory again there is free space.. The allocated BAM is not written back to the disk after a (disk-full) failed save, it is just in the 1541 memory.. I guess we don't want to emulate this behaviour.

dikdom commented:
I think yesterday I found a bug in the 1541 ROM. So, when saving a file that doesn't fit, of course the ST will be set (to -125 or 131) and the disk-write led blinks. But loading the directory it says there is no free space. After initializing the disk (or removing and putting it back) and loading the directory again there is free space.. The allocated BAM is not written back to the disk after a (disk-full) failed save, it is just in the 1541 memory. :)
I will skip this feature of the 1541.

dikdom commented:
Should be addressed all your earlier concerns and suggestions. Please, have a look!

KimJorgensen commented:
Thank you for the changes.
Unfortunately it seems that the “big thing” that I wrote about in my very first comment hasn’t really been addressed other than merging the d64_reader and d64_writer files.
In my last comment (https://github.com/KimJorgensen/KungFuFlash/pull/30#issuecomment-678760019) I described the responsibility of the d64_reader and disk_drive and had hoped that would explain how writing to disk could be implemented following the existing design. But I guess I didn’t do a good job of explaining this so do you have any questions to what I mean?

dikdom commented:
Well yes, then I do. 
I thought you wanted to have one file for D64 management. Originally it was the D64 reader as it had all the D64 read related stuff. Two new functions were added regarding to D64 write and I renamed to d64.c. The other helper functions for save are left in disk_drive.c 
I'm sorry, this is what I thought you were requesting, and actually it made sense.
Anyway, let's go back a few steps. What did I do wrong (please give examples) and how to fix them (guides).

KimJorgensen commented:
Well, I think one d64.c file is fine but the API is not.
It should be easy to write a d64_write_prg function similar to d64_read_prg in d64.c and disk_drive.c should not know about the inner workings of a D64 image. So disk_drive.c should not know anything about the track layout or how the BAM is constructed.

dikdom commented:
... and d64.c routines shall not know anything about the mcu-c64 communication, right?

KimJorgensen commented:
Exactly. That is the responsibility of disk_drive.c

dikdom commented:
Currently I'm struggling with a mysterious bug: during the SAVE procedure when I call in the middle of the procedure (not the first time) the ```d64_read_disk_header(d64)``` function, the C64 crashes (like R/S + RESTORE, just no READY and blinking cursor). There is probably a leaking data structure, but I couldn't locate it yet.

On the other hand, I realized yesterday that the interrupt flag is cleared (interrupts are enabled) during the kernal print call (e.g. when printing 'SEARCHING FOR' or 'SAVING', etc...). Was it the intended way of behavior during the LOAD?

dikdom commented:
@KimJorgensen ready for the next round..

KimJorgensen commented:
Hi,
Having reviewed your latest changes I can see that it does move most of the responsibility for handling disk images to d64.c which I asked for. However, I think that the write API could be simpler and have fewer methods that you need to call and there are a code duplication in both in disk_drive and d64 along with a few other issues. 

Would it be OK with you if I merge you PR to a branch make changes there myself before putting it on the master branch? I don’t find the current process very effective when dealing with such a large PR.

dikdom commented:
Hello,
Sure, no problem at all!

KimJorgensen commented:
Great, thank you!
I have merged it now so you don't have to do any changes
